File: gcode2dplotterart/SimpleInstruction.py
from abc import ABC


class _AbstractSimpleInstruction(
    ABC,
):
    """Abstract class representing an instruction in G-Code. Used for simple instructions that do not require
    user inputs, such as "M2".

    Args:
      instruction (str) : The instruction to use.
      comment (str) : A comment representing the instruction's purpose."""

    def __init__(self, instruction: str, comment: str):
        self.instruction = instruction
        self.comment = comment

    def __str__(self) -> str:
        return f"{self.comment}"

    def to_g_code(self) -> str:
        """
        Convert instruction to G-Code.

        Returns:
          string
            A special instruction in G-Code format.
        """
        return f"{self.instruction}"


class Instruction2DPlotterPlottingHeight(_AbstractSimpleInstruction):
    """
    Connect plotting instrument to plotting surface
    """

    def __init__(self) -> None:
        super().__init__("M3 S1000", "Connect plotting instrument to plotting surface")


class Instruction2DPlotterNavigationHeight(_AbstractSimpleInstruction):
    """
    Separate plotting instrument from plotting surface
    """

    def __init__(self) -> None:
        super().__init__(
            "M3 S0",
            "Separate plotting instrument from plotting surface",
        )


class InstructionUnitsMM(_AbstractSimpleInstruction):
    """
    Set the units of the layer to mm
    """

    def __init__(self) -> None:
        super().__init__("G21", "Set the units of the layer to mm")


class InstructionProgramEnd(_AbstractSimpleInstruction):
    """
    Instruct the plotting device that plotting has completed
    """

    def __init__(self) -> None:
        super().__init__(
            "M2", "Instruct the plotting device that plotting has completed"
        )

File: gcode2dplotterart-ts/src/engine/instructions/SimpleInstruction.ts
export class Instruction2DPlotterPlottingHeight {
  toGCode(): string {
    return "M3 S1000";
  }
}

export class Instruction2DPlotterNavigationHeight {
  toGCode(): string {
    return "M3 S0";
  }
}

export class InstructionUnitsMM {
  toGCode(): string {
    return "G21";
  }
}

export class InstructionHome {
  toGCode(): string {
    return "G28";
  }
}

export class InstructionProgramEnd {
  toGCode(): string {
    return "M2";
  }
}
